x-pack/plugin/core/src/test/java/org/elasticsearch/license/internal/TrialLicenseEraTests.java

```diff
@@ -0,0 +1,98 @@
+
+    public void testCanParseAllVersions() {
+        for (var version : Version.getDeclaredVersions(Version.class)) {
+            TrialLicenseEra era = TrialLicenseEra.fromString(version.toString());
+            assertThat((byte) era.asInt(), equalTo(version.major));
+        }
+    }
+
+    public void testVersionWireCompatibility() throws IOException {
+        for (var version : Version.getDeclaredVersions(Version.class)) {
+            versionToEraSerialization(version);
+        }
+
+        for (int i = 2; i < 100; i++) { // old Version class only grows to 99, so that's the limit
+            eraToVersionSerialization(new TrialLicenseEra(i));
+        }
+    }
+
+    // This simulates a node of a version that supports transport eras receiving a message which is still using old-school Versions
+    private TrialLicenseEra versionToEraSerialization(Version version) throws IOException {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.setTransportVersion(TransportVersion.current());
+            Version.writeVersion(version, output);
+            try (
+                StreamInput in = new NamedWriteableAwareStreamInput(
+                    output.bytes().streamInput(),
+                    new NamedWriteableRegistry(Collections.emptyList())
+                )
+            ) {
+                in.setTransportVersion(TransportVersion.current());
+                TrialLicenseEra readEra = new TrialLicenseEra(in);
+                assertEquals(version.major, (byte) readEra.asInt());
+                return readEra;
+            }
+        }
+    }
+
+    // This simulates (imperfectly) a node of a version that does not support transport eras receiving a message which uses trial version
+    // eras
+    private Version eraToVersionSerialization(TrialLicenseEra era) throws IOException {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.setTransportVersion(TransportVersion.current());
+            era.writeTo(output);
+            try (
+                StreamInput in = new NamedWriteableAwareStreamInput(
+                    output.bytes().streamInput(),
+                    new NamedWriteableRegistry(Collections.emptyList())
+                )
+            ) {
+                in.setTransportVersion(TransportVersion.current());
+                Version readVersion = Version.readVersion(in);
+                assertEquals(readVersion.major, (byte) era.asInt());
+                return readVersion;
+            }
+        }
+    }
+
+    public void testRoundTripParsing() {
+        var randomEra = new TrialLicenseEra(randomNonNegativeInt());
+        assertThat(TrialLicenseEra.fromString(randomEra.toString()), equalTo(randomEra));
+    }
+
+    public void testVersionCanParseAllEras() {
+        for (int i = 2; i <= TrialLicenseEra.CURRENT.asInt(); i++) {
+            Version.fromString(new TrialLicenseEra(i).asVersionString());
```

Comment:
I can remove this test, I found it necessary to gain enough confidence in this solution to be comfortable PR'ing it but it might not be valuable going forward. We might need to maintain the string format, if for some reason we need old versions of ES to be able to read the XContent emitted by newer ones (see the other thread on that point).